fixes #1462, 'secrets' template function issue on Vault kv-v2 secrets engines

yilmazo opened the pull request:
Changed `secrets` template function behavior for Vault kv-v2 secrets engines. Added behavior similar to [vault cli kv list command](https://github.com/hashicorp/vault/blob/master/command/kv_list.go#L86), which adds `/metadata/` to secret path if secrets engine version is 2. 

Fixes #1275 
Fixes #1274

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul-template?pullRequest=1468)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul-template?pullRequest=1468)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul-template?pullRequest=1468).</sub>

Dentrax commented:
`in.tpl`:
```tpl
{{ range secrets "secret/" }}
{{ with secret (printf "secret/%s" .) }}{{ range $k, $v := .Data.data }}
{{ $k }}: {{ $v }}
{{ end }}{{ end }}{{ end }}
```

Reproduce:
```
$ vault kv put secret/foo qux=baz
$ vault kv put secret/bar quux=2
$ go run . -template "in.tpl:out.txt" -vault-renew-token=false -vault-token='root' -once
```

`out.txt`
```bash


qux: 1


quux: 2


```

Thanks @yilmazo, LGTM!

eikenb commented:
Hey @yilmazo, thanks for the PR.

Looks good after a quick once-over but I am curious about your posted example use... the out.txt is listed as having `qux: 1` where you're input showed `$ vault kv put secret/foo qux=baz`. Was that a cut-n-paste error?

Dentrax commented:
> Hey @yilmazo, thanks for the PR.
> 
> Looks good after a quick once-over but I am curious about your posted example use... the out.txt is listed as having `qux: 1` where you're input showed `$ vault kv put secret/foo qux=baz`. Was that a cut-n-paste error?

Sorry, yes it was a cut-n-paste error.

```
$ vault kv put secret/foo qux=1
$ vault kv put secret/bar quux=2
$ go run . -template "in.tpl:out.txt" -vault-renew-token=false -vault-token='root' -once
```

`out.txt`

```bash
quux: 2
qux: 1
```

eikenb commented:
Did you mean to include the changes to vault_write.go [1]? It seems out of scope with the other parts of the PR, but seems to do something useful (presumably better v2 compatibility with write). If you did intend to add it, seems like it could use a test to validate those changes.

[1] https://github.com/hashicorp/consul-template/pull/1468/files#diff-3764290d54697653752a0700e9aaf6f25343d764c550de67a72e281db471774d

yilmazo commented:
> Did you mean to include the changes to vault_write.go [1]? It seems out of scope with the other parts of the PR, but seems to do something useful (presumably better v2 compatibility with write). If you did intend to add it, would you consider writing a test to validate those changes?
> 
> Thanks.
> 
> [1] https://github.com/hashicorp/consul-template/pull/1468/files#diff-3764290d54697653752a0700e9aaf6f25343d764c550de67a72e281db471774d
> 
> [edit: tone of last sentense seemed off, fixed it]

yes, I intentionally made those changes for making VaultWriteQuery work on v2 engines when no `/data/` in secret path. 
I forgot to add test for that change as you mentioned. Just added it 🙂 [1]  

I could also make no changes to vault_write.go and add `data/` prefix for secret path manually when writing secret to v2 on vault_list_test.go [2] But I decided to go for that change because I've seen Vault cli `kv put` command has similar behavior. [3]  

So, do you think it is OK or should I remove those changes on vault_write.go ?  

Thanks.

[1] https://github.com/hashicorp/consul-template/pull/1468/files#diff-391a41398e4903918dae1ee172ba4ba336ae3c57b26dca5d51818a0ca4493049R147
[2] https://github.com/hashicorp/consul-template/pull/1468/files#diff-fba571a697966b4ab7a40eb36688ac7f70ed6a7328c2c5c793442113acbbc106R85
[3] https://github.com/hashicorp/vault/blob/master/command/kv_put.go#L133

eikenb commented:
I'd like to keep the vault_write changes as they do improve that method. I just wasn't 100% if you meant to have it there or forgot the test. 

Thanks for adding the test!
